hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/upgrade/EightToNineUpgradeHandler.java

```diff
@@ -19,55 +19,198 @@
+                                                                         String instantTime,
+                                                                         SupportsUpgradeDowngrade upgradeDowngradeHelper) {
     final HoodieTable table = upgradeDowngradeHelper.getTable(config, context);
-    // Rollback and run compaction in one step
-    rollbackFailedWritesAndCompact(
-        table, context, config, upgradeDowngradeHelper,
-        HoodieTableType.MERGE_ON_READ.equals(table.getMetaClient().getTableType()),
-        HoodieTableVersion.NINE);
-
-    // If auto upgrade is disabled, set writer version to 8 and return
+    Map<ConfigProperty, String> tablePropsToAdd = new HashMap<>();
+    // If auto upgrade is disabled, set writer version to 8 and return.
     if (!config.autoUpgrade()) {
-      config.setValue(HoodieWriteConfig.WRITE_TABLE_VERSION, String.valueOf(HoodieTableVersion.EIGHT.versionCode()));
-      return Collections.emptyMap();
+      config.setValue(
+          HoodieWriteConfig.WRITE_TABLE_VERSION,
+          String.valueOf(HoodieTableVersion.EIGHT.versionCode()));
+      return Pair.of(tablePropsToAdd, Collections.emptyList());
     }
     HoodieTableMetaClient metaClient = table.getMetaClient();
-
+    HoodieTableConfig tableConfig = metaClient.getTableConfig();
     // Populate missing index versions indexes
     Option<HoodieIndexMetadata> indexMetadataOpt = metaClient.getIndexMetadata();
     if (indexMetadataOpt.isPresent()) {
       populateIndexVersionIfMissing(indexMetadataOpt);
-
       // Write the updated index metadata back to storage
       HoodieTableMetaClient.writeIndexMetadataToStorage(
           metaClient.getStorage(),
           metaClient.getIndexDefinitionPath(),
           indexMetadataOpt.get(),
           metaClient.getTableConfig().getTableVersion());
     }
-    return Collections.emptyMap();
+    // If metadata is enabled for the data table, and
+    // existing metadata table is behind the data table, then delete it.
+    checkAndHandleMetadataTable(context, table, config, metaClient);
+    // Rollback and run compaction in one step.
+    rollbackFailedWritesAndCompact(
+        table, context, config, upgradeDowngradeHelper,
+        HoodieTableType.MERGE_ON_READ.equals(table.getMetaClient().getTableType()),
+        HoodieTableVersion.EIGHT);
+    // Handle merge mode config.
+    upgradeMergeModeConfig(tablePropsToAdd, tableConfig);
+    // Handle partial update mode config.
+    upgradePartialUpdateModeConfig(tablePropsToAdd, tableConfig);
+    // Handle merge properties config.
+    upgradeMergePropertiesConfig(tablePropsToAdd, tableConfig);
+    // Handle payload class configs.
+    List<ConfigProperty> tablePropsToRemove = new ArrayList<>();
+    upgradePayloadClassConfig(tablePropsToAdd, tablePropsToRemove, tableConfig);
+    return Pair.of(tablePropsToAdd, tablePropsToRemove);
+  }
+
+  private void upgradePayloadClassConfig(Map<ConfigProperty, String> tablePropsToAdd,
```

Comment:
the `upgradeXXXX` naming is misleading. it feels as though you are actually changing storage.. 

rename all these to `changeXXX` e.g `changePayloadClassConfigs`